pinot-core/src/main/java/org/apache/pinot/core/transport/QueryRouter.java

```diff
@@ -60,20 +84,23 @@ public AsyncQueryResponse submitQuery(long requestId, String rawTableName,
       long timeoutMs) {
     assert offlineBrokerRequest != null || realtimeBrokerRequest != null;
 
+    // can prefer but not require TLS until all servers guaranteed to be on TLS
```

Comment:
Not sure I understand completely. If I follow the code, it seems like preferTLS == requireTLS. If it errors out, do we try to go with plain? Maybe I missed that code?

Reply:
prefers != requires. the call to `toServerRoutingInstance()` below will attempt to retrieve a ServerRoutingInstance with `tlsEnabled == true` but may not be successful in doing so (if the server doesn't offer netty-tls yet). In that case, we connect via unsecured netty (see QueryRouter:115)